Advanced backend configuration via JSON

franzpoeschel opened the pull request:
This PR adds backend-specific configuration options. Close #150. The configuration is passed as an optional string parameter in JSON format to the constructor of `openPMD::Series`.
The top-level JSON maps backends to the corresponding backend's configuration. This allows to configure several backends at once.

At the moment, only the ADIOS2 backend has been made aware of this. A possible configuration looks like this:
```cpp
std::string options = R"(
{
  "adios2": {
    "engine": {
      "type": "sst",
      "parameters": ["BufferGrowthFactor": "2.0", "QueueLimit": "2"]
    }
  }
}
)";
```
TODO:


- [x] Safety measure: Warn upon reading an unexpected key.
- [x] Add Python bindings
- [x] Code cleanup
- [x] Add user documentation, e.g. in `docs/source/backends/adios2.rst`

Probably follow-up for a later PR:

- [ ] Implement the suggestions in issue #688 
- [ ] Further configuration options in ADIOS2, e.g. switching on and off debug mode at runtime.
- [ ] Implement option reading in other backends
- [ ] What about case insensitivity? Should the JSON keys that are read by the openPMD API and not directly passed through to the backend (such as the ADIOS2 "parameters" list) be made case insensitive?
- [ ] Override choice of backend from JSON. Choosing the backend by filename extension is simple, yet inflexible if one backend has several possible extensions (ADIOS2) or two backens share an extension (ADIOS in versions 1 and 2). Just overwrite setting introduced in #568

ax3l commented:
I updated the CI with further ADIOS2 tests in #568.
Please rebase against the current `dev` :)

ax3l commented on include/openPMD/IO/AbstractIOHandlerHelper.hpp:
```diff
@@ -36,26 +36,28 @@ namespace openPMD
+ * @param   accessType  AccessType describing desired operations and permissions
+ * of the desired handler.
+ * @param   format      Format describing the IO backend of the desired handler.
+ * @return  Smart pointer to created IOHandler.
```
`options` param is still missing :)

ax3l commented on include/openPMD/IO/AbstractIOHandlerHelper.hpp:
```diff
@@ -36,26 +36,28 @@ namespace openPMD
      * @param   comm        MPI communicator used for IO.
      * @return  Smart pointer to created IOHandler.
```
`options` param is still missing :)

ax3l commented on include/openPMD/Series.hpp:
```diff
@@ -226,19 +234,21 @@ class Series : public Attributable
      *
-     * @param   name    String of the pattern for file names. Must include iteration regex <CODE>\%T</CODE> for fileBased data.
+     * @param   name    String of the pattern for file names. Must include
+     * iteration regex <CODE>\%T</CODE> for fileBased data.
```
Maybe align this to beginning of description?

ax3l commented on include/openPMD/Series.hpp:
```diff
@@ -226,19 +234,21 @@ class Series : public Attributable
 
-OPENPMD_private:
-    struct ParsedInput;
+    OPENPMD_private : struct ParsedInput;
```
this syntax confuses me, probably an auto-formatting issue?

ax3l commented on src/IO/AbstractIOHandlerHelper.cpp:
```diff
@@ -18,66 +18,73 @@
 #include "openPMD/IO/AbstractIOHandlerHelper.hpp"
-#include "openPMD/IO/DummyIOHandler.hpp"
+
+#include <openPMD/IO/ADIOS/ADIOS2IOHandler.hpp>
```
```suggestion
#include "openPMD/IO/ADIOS/ADIOS2IOHandler.hpp"
```
and no extra empty line

ax3l commented on src/IO/AbstractIOHandlerHelper.cpp:
```diff
@@ -18,66 +18,73 @@
 #include "openPMD/IO/HDF5/HDF5IOHandler.hpp"
 #include "openPMD/IO/HDF5/ParallelHDF5IOHandler.hpp"
 #include "openPMD/IO/JSON/JSONIOHandler.hpp"
+#include <nlohmann/json.hpp>
```
extra empty line

ax3l commented:
@franzpoeschel I rebased this for you.

ax3l committed:
Gracefully handle missing JSON

JSON is still optional, so we should gracefully throw if a user tries
to pass options that we cannot parse without it.

https://travis-ci.org/openPMD/openPMD-api/builds/599388433

ax3l commented:
We should decide if we want to make `<nlohmann/json.hpp>` a non-optional dependency, which looks potentially solid enough to me to make this call, or if we want to cleanly error out if a user tries to pass runtime options without having JSON available.

I will commit a patch for the latter for now, but it's not pretty.

ax3l commented:
I implemented the selection of the preferred `BP` backend via an environment variable here:
https://github.com/openPMD/openPMD-api/blob/731a4073bf5661da853001bb54ad13709a38c18d/src/Series.cpp#L868-L887

We can also overwrite that by a runtime option, like the other options you added.

Adding control to HDF5 for collective/independent IO, via `H5FD_MPIO_INDEPENDENT`, to address #554 would be a good thing to implement right away as well.

ax3l commented:
ping @franzpoeschel any thoughts on this? :)
Would like to make this available for the next release, so I can address the HDF5 control for independent I/O via a runtime API option as well.

franzpoeschel commented:
I'm fine with controlling the `.bp` backend with an environment variable. I think that we should add a parameter to the JSON string to explicitly choose backends that takes precedence over other options.
Making the other backends aware of this configuration style would be good (as you suggested for HDF5), I just don't have the time at the moment.
I think there is no issue with making `<nlohmann/json.jpp>` a non-optional dependency? It is currently not included in user-facing code, and this would also mean that the API will compile with at least one usable backend, simplifying accessibility.

ax3l commented:
> I'm fine with controlling the .bp backend with an environment variable. I think that we should add a parameter to the JSON string to explicitly choose backends that takes precedence over other options.

Agreed, the env var was a quick hack of mine but I want users to be able to control it via an option, which is more friendly and which is why I try to merge this PR soon :)

> Making the other backends aware of this configuration style would be good (as you suggested for HDF5), I just don't have the time at the moment.

Argh, too bad. Can you potentially finish the ADIOS impl. with it and I take a look at the other backends?

> I think there is no issue with making <nlohmann/json.jpp> a non-optional dependency.

Excellent, agreed. Will do that in an independent PR then (#587).

Edit: branch updated and rebased so you can finish the ADIOS2 controls.

ax3l commented:
ping @franzpoeschel finishing this one would also be really nice, please :)

franzpoeschel commented:
My plan is to have topic-streaming go on top of this one, so this is part of the endeavours ;)

ax3l commented:
Sounds great, I already made the JSON backend required so you just need to rebase to continue.

franzpoeschel committed:
Add JSON-formatted string parameter to frontend for backend
configuration

franzpoeschel committed:
Add a simple implementation of JSON parsing to ADIOS2 backend

ax3l commented:
@franzpoeschel just since I need this more urgently for HDF5 option control and compression support: when do you think you can continue to work on this?

franzpoeschel commented:
@ax3l My defense talk is next wednesday, I can fully focus on openPMD again after that :)

ax3l commented:
Wish you all the best for your defense, please feel free to send me your slides on my lbl e-mail if you want me to take a look. Fingers crossed! :crossed_fingers: :muscle:

franzpoeschel committed:
Add TracingJSON class

This will allow warning users of unused entries in the JSON config.

Invert TracingJSON shadow logic

Instead of removing all accessed entries one after another, add them to
the "shadow" value one after another. This is a bit more tricky since
the result has to be inverted again in the end, but it is less prone to
creating dangling pointer situations.

franzpoeschel committed:
Move ADIOS2 backend to TracingJSON

Warn in ADIOS2 when not using all of the JSON configuration

franzpoeschel commented:
For testing purposes, it might be sensible to link the `nlohmann_json::nlohmann_json` target also with the testing targets, so tests that directly use that library can be written as well.

ax3l commented:
I admittedly do not want to introduce this as a public library dependency and would prefer to keep this a (cmake-lingo-)private dep.
The reason behind this is that this would otherwise complicate downstream builds that want to use openPMD-api (e.g. number of to-be-set include paths) for users without CMake toolchains. Also, I want to limit the number of (public) dependencies that have to compile when included in a `nvcc` TU, which is basically everything that goes through the public includes to our users.

But we can certainly provide an auxiliary target in CMake that can be optionally be used.

Anyway, in the tests I would prefer just to write the json options as strings (or nested init lists) for now, to guarantee that the above mentioned aspect works.

franzpoeschel commented:
The reason why I am suggesting this is because with the latest commit, I introduced a [JSON class](https://github.com/openPMD/openPMD-api/blob/5c0086a000c189fa96b66dc4b621e8b15200b8e6/include/openPMD/auxiliary/JSON.hpp) depending on `nlohmann::json` that will trace accesses (in order to give warnings for unused keys). With the current state of the build system, I have only been able to test this externally by including that class and also N. Lohmann's JSON library in another project. Since it's only for warnings and not "mission-critical", I could leave it at that – or do you have another suggestion?

ax3l commented:
Is it possible you did not rebase against the latest `dev` yet? I adjusted the build system already a few months back to make `nlohmann_json` a mandatory, private dependency of openPMD-api #587 

Notes:
- please rebase against the latest `dev` branch
- you don't need `#if openPMD_HAVE_JSON` guards anymore, the `<nlohmann/json.hpp>` include is now always guaranteed to be there by the build system, as long as you build the primary source files of the CMake `openPMD` target, which your auxiliary `JSON.cpp` file is part of.
- as you do already: include `openPMD/auxiliary/JSON.hpp` only in `.cpp` files of the openPMD library target and do not add this include to the facade header `include/openPMD/openPMD.hpp` to avoid a public dependency on `<nlohmann/json.hpp>` for users

franzpoeschel committed:
Expose Python bindings for JSON configuration

franzpoeschel committed:
Make TracingJSON more efficient

operator[] will no longer copy the whole JSON structure

Make TracingJSON more efficient

operator[] will no longer copy the whole JSON structure

franzpoeschel committed:
Add user documentation for backend-specific configuration

Update ADIOS2 configuration documentation

franzpoeschel committed:
Fix CI errors

name shadowing and documentation

franzpoeschel committed:
Some cleanup in ADIOS2IOHandlerImpl

franzpoeschel committed:
Enable ADIOS2 operator reading from JSON config

Also fix a bug where a null pointer is dereferenced if an invalid
compression is requested.

ax3l commented:
@franzpoeschel when this is ready for review: please remove the `[WIP]` tag from the PR description.

Also: can you add a small test, e.g. using compression, setting an engine, etc.?

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+.. _backendconfig
+
+Backend-specific configuration
```
English titles:
```suggestion
Backend-Specific Configuration
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+Backend-specific configuration
+==============================
+
+While the openPMD API intends to be a backend-*independent* implementation of the openPMD standard, it is often necessary to pass configuration parameters to the concrete backend in use. Some backends recognize a number of configuration options via environment variables (see the backends' documentations), the preferred way to do so, however, *– as far as it is implemented –* is via a JSON-formatt ...
```
```suggestion
While the openPMD API intends to be a backend-*independent* implementation of the openPMD standard, it is sometimes useful to to pass configuration parameters to the specific backend in use.
:ref:`For each backend <backends-overview>`, configuration options can be passed via a JSON-formatted string or via environment variables.
A JSON option always takes precedence over an environment variable.
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+
+This structure allows keeping one configuration string for several backends at once, with the concrete backend configuration being chosen upon choosing the backend itself.
+
+The configuration is currently read in a case-sensitive manner. Since this may change in the future, all parts of the configuration are given in *lower case*. Parameters that are directly passed through to an external library and not interpreted within the openPMD API (e.g. ``adios2.engine.parameters``) are unaffected by this and follow the respective library's conventions.
```
One sentence per line please:
```suggestion
The configuration is read in a case-sensitive manner.
Generally, keys of the configuration are *lower case*.
Parameters that are directly passed through to an external library and not interpreted within openPMD API (e.g. ``adios2.engine.parameters``) are unaffected by this and follow the respective library's conventions.
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+
+The configuration is currently read in a case-sensitive manner. Since this may change in the future, all parts of the configuration are given in *lower case*. Parameters that are directly passed through to an external library and not interpreted within the openPMD API (e.g. ``adios2.engine.parameters``) are unaffected by this and follow the respective library's conventions.
+
+The configuration string may refer to the complete ``openPMD::Series`` and additionally be specified per ``openPMD::Dataset``, passed in the respective constructors. *Configuration per dataset is currently not yet implemented.* This reflects the fact that certain backend-specific parameters may refer to the whole Series (such as storage engines and their parameters) and others refer to actual dat ...
```
```suggestion
The configuration string may refer to the complete ``openPMD::Series`` or may additionally be specified per ``openPMD::Dataset``, passed in the respective constructors.
*A configuration per dataset is currently not yet implemented.*
This reflects the fact that certain backend-specific parameters may refer to the whole Series (such as storage engines and their parameters) and others refer to actual datasets (such as compression).
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+* The configuration structures for the Series and for each dataset should be defined equivalently.
+* Any setting referring to single datasets should also be applicable globally, affecting all datasets.
+* If a setting is defined globally, but also for a concrete dataset, the dataset-specific setting should override the global one.
+* If a setting is passed to a dataset that only makes sense globally (such as the storage engine), the setting should be ignored except for printing a warning. Backends should define clearly which keys are applicable to datasets and which are not.
```
```suggestion
* If a setting is passed to a dataset that only makes sense globally (such as the storage engine), the setting should be ignored except for printing a warning.
  Backends should define clearly which keys are applicable to datasets and which are not.
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+* If a setting is defined globally, but also for a concrete dataset, the dataset-specific setting should override the global one.
+* If a setting is passed to a dataset that only makes sense globally (such as the storage engine), the setting should be ignored except for printing a warning. Backends should define clearly which keys are applicable to datasets and which are not.
+
+Configuration structure per backend
```
```suggestion
Configuration Structure per Backend
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,47 @@
+
+The fundamental structure of this JSON configuration string is given as follows:
+
+.. literalinclude:: config_layout.json
```
Follow-up PR: can we add a [schema](https://json-schema.org/) for this?

franzpoeschel replied:
Probably a good idea, yeah

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -258,6 +298,14 @@ class ADIOS2IOHandlerImpl
+     * The following strings are used during parsing of the JSON configuration
+     * string for the ADIOS2 backend.
+     */
+    constexpr char const * const str_engine = "engine";
```
global vars: :scream: 

can we place them in a struct or so? e.g. `struct ADIOS2defaults { ... };`?

franzpoeschel replied:
They are `constexpr`, so not really variables ;)
The alternative would be using `#define` or using literals directly in the code, I think this is the cleanest solution. Putting them in a struct is a good idea, will do that.

ax3l commented on include/openPMD/auxiliary/JSON.hpp:
```diff
@@ -0,0 +1,116 @@
+
+#include "openPMD/config.hpp"
+
+#if openPMD_HAVE_JSON
```
see previous comments: `openPMD_HAVE_JSON` guards are not used in the code anymore.

ax3l commented on include/openPMD/auxiliary/JSON.hpp:
```diff
@@ -0,0 +1,116 @@
+} // namespace auxiliary
+} // namespace openPMD
+
+#endif // openPMD_HAVE_JSON
```
missing EOF newline

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -20,18 +20,21 @@
 #include "openPMD/IO/ADIOS/ADIOS2IOHandler.hpp"
 #include "openPMD/auxiliary/Environment.hpp"
 #include "openPMD/auxiliary/Filesystem.hpp"
+// #include "openPMD/auxiliary/JSON.hpp"
```
dead code?

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -58,25 +61,59 @@ namespace openPMD
+    )
     : AbstractIOHandlerImplCommon( handler ), m_ADIOS{ ADIOS2_DEBUG_MODE }
 {
+#    if openPMD_HAVE_JSON
```
see above

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -515,6 +551,10 @@ ADIOS2IOHandlerImpl::adios2Accesstype()
     }
 }
 
+#    if openPMD_HAVE_JSON
```
see above

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -1116,58 +1157,114 @@ namespace detail
+    {
+        (void)impl;
+        std::set< std::string > alreadyConfigured;
+#    if openPMD_HAVE_JSON
```
see above

franzpoeschel committed:
Add reviewer's comments concerning documentation

franzpoeschel committed:
Remove openPMD_HAVE_JSON macros

JSON is now a mandatory dependency

franzpoeschel committed:
Move ADIOS2 defaults to struct of their own

franzpoeschel committed:
Add JSON config test

franzpoeschel commented:
In the [test for this PR](https://github.com/openPMD/openPMD-api/blob/28ebb36f64ec0d7f449db69dd0c9d5e7fbf11ab2/test/SerialIOTest.cpp#L2192), I have not been able to catch an expected exception thrown by ADIOS2, neither with `REQUIRE_THROWS`, nor with `try..catch`. Might this be related to [this issue](https://stackoverflow.com/questions/38878999/try-catch-doesnt-work-in-shared-library)? 
EDIT: Recompiling with `-shared-libgcc` did not change this, the exception is not caught. Output from `ctest`:
```
3: terminate called after throwing an instance of 'std::ios_base::failure[abi:cxx11]'
3:   what():  ERROR: couldn't read from file ../samples/jsonConfigured.bp, in call to POSIX IO read                                                                                           
3: : iostream error         
3:                                             
3: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
3: SerialIOTests is a Catch v2.6.1 host application.                                           
3: Run with -? for options
3:   
3: -------------------------------------------------------------------------------            
3: serial_adios2_json_config
3: -------------------------------------------------------------------------------             
3: /home/franz/workspace/openPMD-api/test/SerialIOTest.cpp:2149                               
3: ...............................................................................             
3:   
3: /home/franz/workspace/openPMD-api/test/SerialIOTest.cpp:2149: FAILED:                      
3: due to a fatal error condition:                                                             
3:   SIGABRT - Abort (abnormal termination) signal                                            
3:                                                                                             
3: ===============================================================================            
3: test cases:  14 |  13 passed | 1 failed                                                     
3: assertions: 487 | 486 passed | 1 failed                                                     
3:                                                                                             
3/4 Test #3: Serial.SerialIO ..................Child aborted***Exception:   0.24 sec
```

I have deactivated the corresponding check for now.

guj commented:
Hi Franz,
It is due to ADIOS2IOHandler::flush( )
returns a future object.

Would future be needed here?

franzpoeschel committed:
Cleanup operator handling

franzpoeschel commented:
Hi @guj,
`ADIOS2IOHandler::flush` returns a `std::future` since it overrides `AbstractIOHandlerImpl::flush` which also does, in order to support future backends that may work asynchronously. The ADIOS2 backend does not make use of that and returns the default-constructed `std::future` created in [`AbstractIOHandlerImpl.hpp`](https://github.com/openPMD/openPMD-api/blob/c16ffd22993027eef46281724c7fa920f7851ca0/include/openPMD/IO/AbstractIOHandlerImpl.hpp#L115).

If I understand you correctly, you are suggesting that the exception is thrown in asynchronous code (i.e. inside the `std::future`) which I verified is not the case. I also tried catching the exception directly upon opening the engine in [`detail::BufferedActions::getEngine`](https://github.com/openPMD/openPMD-api/blob/c16ffd22993027eef46281724c7fa920f7851ca0/src/IO/ADIOS/ADIOS2IOHandler.cpp#L1174) which did not work either.

I think I will simply rewrite the test to not rely on throwing and catching exceptions for now.

franzpoeschel committed:
Check whether files have been written with correct ADIOS2 engine

JSON configuration test will verify that the configuration of the ADIOS2
engine has actually had an effect.

franzpoeschel committed:
Add EOF newline to JSON.hpp

franzpoeschel commented:
I fixed the test now.

Does the CI currently have some issues?
```
HDF5-DIAG: Error detected in HDF5 (1.10.1) thread 0:

  #000: H5F.c line 586 in H5Fopen(): unable to open file

    major: File accessibilty

    minor: Unable to open file

  #001: H5Fint.c line 1236 in H5F_open(): unable to open file: time = Wed Mar  4 14:34:08 2020

, name = '../samples/git-sample/data00000100.h5', tent_flags = 1

    major: File accessibilty

    minor: Unable to open file

  #002: H5FD.c line 809 in H5FD_open(): open failed

    major: Virtual File Layer

    minor: Unable to initialize object

  #003: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = '../samples/git-sample/data00000100.h5', errno = 13, error message = 'Permission denied', flags = 1, o_flags = 2

    major: File accessibilty

    minor: Unable to open file

git sample not accessible. ([HDF5] Failed to open HDF5 file ../samples/git-sample/data00000100.h5)
```
(https://travis-ci.org/openPMD/openPMD-api/jobs/658223474)

ax3l commented on src/IO/AbstractIOHandlerHelper.cpp:
```diff
@@ -36,9 +39,15 @@ namespace openPMD
+        MPI_Comm comm,
+        std::string const & options )
     {
+#   if openPMD_HAVE_JSON
```
we always have `openPMD_HAVE_JSON`

franzpoeschel replied:
All traces of this should be removed from the source tree now, according to `grep`.

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,55 @@
+The fundamental structure of this JSON configuration string is given as follows:
+
+.. literalinclude:: config_layout.json
+
```
Not sure if this will be already nicely highlighted, otherwise add:
```suggestion
   :language: json
   
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,55 @@
+.. _backendconfig
+
+Backend-Specific configuration
```
Titles:
```suggestion
Backend-Specific Configuration
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,55 @@
+A full configuration of the ADIOS2 backend:
+
+.. literalinclude:: adios2.json
+
```
See above, if needed:
```suggestion
   :language: json
   
```

franzpoeschel replied:
I already applied the same fix to the documentation of the JSON backend.

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,55 @@
+
+.. literalinclude:: adios2.json
+
+All keys found under ``adios2.dataset`` are applicable globally as well as per dataset, keys found under ``adios2.engine`` only globally. Explanation of the single keys:
```
One sentence per line:
```suggestion
All keys found under ``adios2.dataset`` are applicable globally as well as per dataset, keys found under ``adios2.engine`` only globally.
Explanation of the single keys:
```

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -48,6 +38,17 @@
+#include <utility> // pair
+#include <vector>
+
+#include "openPMD/auxiliary/JSON.hpp"
```
this should be in the first block of includes

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -48,6 +38,17 @@
+#include <vector>
+
+#include "openPMD/auxiliary/JSON.hpp"
+#include <nlohmann/json.hpp>
```
this should be before the std lib includes

franzpoeschel replied:
Is it [better now](https://github.com/openPMD/openPMD-api/blob/61be5f724229a056da50c1959519016eb3f7fbc2/include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp#L23)?

ax3l replied:
Not 100% yet. I would
- prefix `ADIOS2FilePosition.hpp` with the usual full path
- include `openPMD/IO/ADIOS/ADIOS2Auxiliary.hpp` before third party libs

```C++
#include "openPMD/config.hpp"
#include "openPMD/IO/AbstractIOHandler.hpp"
#include "openPMD/IO/AbstractIOHandlerImpl.hpp"
#include "openPMD/IO/AbstractIOHandlerImplCommon.hpp"
#include "openPMD/IO/IOTask.hpp"
#include "openPMD/IO/InvalidatableFile.hpp"
#include "openPMD/IO/ADIOS/ADIOS2FilePosition.hpp"
#include "openPMD/auxiliary/JSON.hpp"
#include "openPMD/backend/Writable.hpp"

#if openPMD_HAVE_ADIOS2
#   include "openPMD/IO/ADIOS/ADIOS2Auxiliary.hpp"
#   include <adios2.h>
#endif

#include <nlohmann/json.hpp>

#if openPMD_HAVE_MPI
#   include <mpi.h>
#endif

#include <array>
// ...
```

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -256,6 +308,20 @@ class ADIOS2IOHandlerImpl
+ * Note that this struct's members additionally need to be defined at namespace
+ * scope up until C++17, see ADIOS2IOHandlerImpl.cpp
+ */
+struct ADIOS2Defaults
```
Is there an advantage by doing this in a struct with `static` members over just taking a `namespace` with the same name and avoiding all the `static` handling?

franzpoeschel replied:
Yeah, using a namespace is better. Fixed it.

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -472,10 +534,15 @@ namespace detail
-                        std::unique_ptr< adios2::Operator > compression,
-                        const adios2::Dims & shape, const adios2::Dims & start,
-                        const adios2::Dims & count, bool constantDims );
+        defineVariable(
```
Maybe add a doxygen parameter description here?

franzpoeschel replied:
done

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -990,11 +1098,12 @@ namespace detail
-            var.AddOperation( *compression );
+            if( compression.op )
+            {
+                var.AddOperation( compression.op, compression.params );
```
Just to make sure this works:
How would I set parameters of blosc, e.g.
```
    // add operations
    adios2::Operator BloscOp =
        adios.DefineOperator("BloscCompressor", adios2::ops::LosslessBlosc);

    varBig.AddOperation(BloscOp,
                         {{adios2::ops::blosc::key::compressor, "zstd"}});
    varBig.AddOperation(BloscOp,
                         {{adios2::ops::blosc::key::clevel, "1"}});
    varBig.AddOperation(BloscOp,
                         {{adios2::ops::blosc::key::doshuffle, "BLOSC_BITSHUFFLE"}});
```

via json?

franzpoeschel replied:
I updated [the test](https://github.com/openPMD/openPMD-api/blob/f21074558c308398e6dda2cca3507a423db95ee0/test/SerialIOTest.cpp#L2166) accordingly. I skipped setting the compressor to `zstd` since ADIOS2 will fail if that is not available.

ax3l replied:
Okay... but it's our default compressor we use all the time in production. Maybe add it in some way to tests nonetheless...

@chuckatkins @guj is it possible to query the ADIOS2 API at runtime to be sure that e.g. blosc was compiled in?

guj replied:
Hi Axel,  Chuck will add it in.

ax3l replied:
@guj do we already have an issue or PR to link here, for reference?

guj commented:
> Hi @guj,
> `ADIOS2IOHandler::flush` returns a `std::future` since it overrides `AbstractIOHandlerImpl::flush` which also does, in order to support future backends that may work asynchronously. The ADIOS2 backend does not make use of that and returns the default-constructed `std::future` created in [`AbstractIOHandlerImpl.hpp`](https://github.com/openPMD/openPMD-api/blob/c16ffd22993027eef46281724c7fa920f7851ca0/include/openPMD/IO/AbstractIOHandlerImpl.hpp#L115).
> 
> If I understand you correctly, you are suggesting that the exception is thrown in asynchronous code (i.e. inside the `std::future`) which I verified is not the case. I also tried catching the exception directly upon opening the engine in [`detail::BufferedActions::getEngine`](https://github.com/openPMD/openPMD-api/blob/c16ffd22993027eef46281724c7fa920f7851ca0/src/IO/ADIOS/ADIOS2IOHandler.cpp#L1174) which did not work either.
> 
> I think I will simply rewrite the test to not rely on throwing and catching exceptions for now.

Hi Franz, yes you are right std::future is not the cause of this.  Especially ADIOS2IOHandler is not really doing async operation. Thanks for double checking.  I took a closer look today. The issue is still related to the flush() function.   It turns out when exception is thrown in flush(), it delegates and eventually the smart pointers are cleaning up the memory. When the ADIOS2IOHandler destructor is called,  the flush() is invoked again.  This time the same exception is thrown again and no where to go. One quick workaround is to do a try and catch in the destructor of ADIOS2IOHandler.  I did a quick test(reading a series from non existing bp file) with this workaround  and the exception from adios2 is then caught by my test.  Hope this makes sense.

franzpoeschel committed:
Extend tests to include adios operator parameters

franzpoeschel committed:
Make ADIOS2Defaults a namespace

(was previously a struct)

franzpoeschel committed:
Implement reviewer's comments

Remove traces of openPMD_HAVE_JSON
Improve JSON highlighting in documentation
Formatting improvements in documentation
Includes reordering
Documentation

franzpoeschel committed:
Use nlohmann::json::get to excplicitly get a string

ax3l commented:
CI is up again, ready for rebase against `dev`.

ax3l commented:
@franzpoeschel looks like this PR reproducibly breaks the `clang@5.0.0 +OMPI -PY +H5 +ADIOS1 (BP3) +ADIOS2` HDF5 tests.

franzpoeschel commented:
Yep. I still need to figure out what's up with that.

ax3l commented:
Re: ADIOS2 exception problem.
This is confusing, since I am able to catch shared library exceptions of openPMD-api (shared) in the python bindings (also shared)...
How did you build ADIOS2 and which version? Manually or is this what you observed both locally and in CI? If you can isolate the few lines for a minimal reproducer, we can ask upstream with ADIOS2 as well because its likely build system related.

guj commented:
@ax3l 
I am using shared library,  the master branch (I don't think branch matters). 
Try this simple test in cpp. See if it crashes (It should.) 
Next you put try/catch block, see if it still crashes (will)  

int main(int argc, char *argv[])
{
    MPI_Init(&argc, &argv);
    Series series = Series("no_such_file.bp",  AccessType::READ_ONLY,  MPI_COMM_WORLD);        
}

ax3l commented:
I think the full issue we talk about is more
```C++
#include <openPMD/openPMD.hpp>

int main()
{
    try
    {
        auto series = openPMD::Series(
            "test.bp",
            openPMD::AccessType::READ_ONLY
        );
    } catch( ... ) {}
    return 0;
}
```

which works with H5 (intentional verbose output but no crash) and fails with ADIOS2 (core dump due to missing catch). But I get the point, good demonstrator :)

guj commented:
> I think the full issue we talk about is more
> 
> ```c++
> #include <openPMD/openPMD.hpp>
> 
> int main()
> {
>     try
>     {
>         auto series = openPMD::Series(
>             "test.bp",
>             openPMD::AccessType::READ_ONLY
>         );
>     } catch( ... ) {}
>     return 0;
> }
> ```
> 
> which works with H5 (intentional verbose output but no crash) and fails with ADIOS2 (core dump due to missing catch). But I get the point, good demonstrator :)

Actually, my full argument is (assuming the bp file does not exist) 
1)  Series::Init(),  eventually calls ADIOS2IOHandler::flush()  from readBase(), around line  Series.cpp 782
2)  it calls m_impl.flush().  Exception is thrown. Msg looks like "ERROR: couldn't open file ...". And  currently this exception is not caught.  
3) plus, ADIOS2IOHandler.hpp:647, flush() is called again in the destructor when shared pointer releases resources when test ends. The same exception in 2) will be thrown again. This will not be caught in the test, so catch block is needed in either ADIOS2IOHandler/Impl. I would just add it in the destructor for a quick workaround.

ax3l commented:
Ups, clicked the wrong button.

Ah, the problem is in flush! of course! The destructor must not throw, we have to indeed catch in `Series::~Series`. Will do a PR.

franzpoeschel commented:
[This line](https://github.com/openPMD/openPMD-api/pull/569/commits/e7022771b7fe0c4300b4291251918a392a99b6cd) seems to be the cause for the broken build. Apparently in that build configuration, the BP4 engine will not write a directory? It can still be opened without any issue (the file will be read from in a later line).
Is there some way to inspect the files created by the CI builds?

ax3l commented:
> This line seems to be the cause for the broken build. Apparently in that build configuration, the BP4 engine will not write a directory?

Careful, there is an ADIOS2 bug that does not write an ADIOS2 directory if there is already an old ADIOS1 `.bp` file (BP3) with the same name in place. When testing locally, better disable ADIOS1 to be definitely sure. in CI, we only execute one or the other. I think we have not yet reported this upstream...

>  Is there some way to inspect the files created by the CI builds?

not really, besides adding some temporary `ls`/`cat`/`find` lines to `.travis.yml`

franzpoeschel commented:
From that bug description, this shouldn't really happen during CI builds then?

franzpoeschel committed:
Don't run ADIOS2 configuration test if ADIOS1 is chosen per env-var

Also remove some outdated todos.

franzpoeschel commented:
I think I've found the issue. The test that broke explicitly set `OPENPMD_BP_BACKEND=ADIOS1`, so the JSON string was completely ignored.
EDIT: After fixing, that test passes now. `gcc@6.5.0 -MPI +PY@3.5 +H5 +ADIOS1@1.13.1` fails during `apt-get install`. Should I trigger a new build?

ax3l commented:
Awesome!

> gcc@6.5.0 -MPI +PY@3.5 +H5 +ADIOS1@1.13.1 fails during apt-get install. Should I trigger a new build?

I just re-scheduled on travis the one build that failed (and the coverage test in the stage behind it). Sometimes fetch/install fails on travis due to network issues.

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -0,0 +1,62 @@
+Configuration Structure per Backend
+-----------------------------------
+
+ADIOS2
```
Generally, and I think this can be a follow-up PR as well, I think these backend-specific options should go into the respective backend docs page (where we also document the current environment vars).

ax3l replied:
@franzpoeschel please try to split this up in a follow-up PR, we already have backend-specific docs where this should go :)

ax3l commented on docs/source/details/config_layout.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "adios": "put ADIOS config here",
+    "adios2": "put ADIOS2 config here",
+    "hdf5": "put HDF5 config here"
```
Could be relevant, because nlohmann-json also supports a couple of binary JSON representations that we might want to use:
```suggestion
    "hdf5": "put HDF5 config here",
    "json": "put JSON config here"
```

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -40,6 +39,11 @@
 #include <utility> // pair
 #include <vector>
 
+#include <nlohmann/json.hpp>
```
- `"openPMD/config.hpp"` should be in the first block of includes :)
- json: somehow included 2x :)
- stdlibs: should be included last

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -95,18 +99,22 @@ class ADIOS2IOHandlerImpl
-    explicit ADIOS2IOHandlerImpl( AbstractIOHandler * );
+    explicit ADIOS2IOHandlerImpl(
+        AbstractIOHandler *
+        , nlohmann::json config
+    );
 
 
-    ~ADIOS2IOHandlerImpl( ) override = default;
+    ~ADIOS2IOHandlerImpl() override = default;
 
-    std::future< void > flush( ) override;
+    std::future< void >
+    flush() override;
```
this block has unnecessary whitespace changes and somehow an odd linebreak in the constructor parameters

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -258,6 +313,20 @@ class ADIOS2IOHandlerImpl
+/*
+ * The following strings are used during parsing of the JSON configuration
+ * string for the ADIOS2 backend.
+ * Note that this struct's members additionally need to be defined at namespace
+ * scope up until C++17, see ADIOS2IOHandlerImpl.cpp
+ */
```
Is this `Note` outdated?

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -258,6 +313,20 @@ class ADIOS2IOHandlerImpl
+/*
+ * The following strings are used during parsing of the JSON configuration
+ * string for the ADIOS2 backend.
+ * Note that this struct's members additionally need to be defined at namespace
+ * scope up until C++17, see ADIOS2IOHandlerImpl.cpp
```
Is this `Note` outdated?

ax3l commented on include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp:
```diff
@@ -666,11 +754,16 @@ friend class ADIOS2IOHandlerImpl;
+    ADIOS2IOHandler(
+        std::string path,
+        AccessType,
+        MPI_Comm
+        , nlohmann::json options
```
```suggestion
        MPI_Comm,
        nlohmann::json options
```

ax3l commented on include/openPMD/auxiliary/JSON.hpp:
```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <memory>  // std::shared_ptr
+#include <utility> // std::forward
```
Include order:
- own header (in `.cpp`)
- openPMD-api headers
- third party libs
- std libs

ax3l commented on include/openPMD/auxiliary/JSON.hpp:
```diff
@@ -0,0 +1,112 @@
+        declareFullyRead();
+
+    private:
+        std::shared_ptr< nlohmann::json > m_originalJSON;
+        std::shared_ptr< nlohmann::json > m_shadow;
+        nlohmann::json * m_positionInOriginal;
+        nlohmann::json * m_positionInShadow;
+        bool m_trace = true;
```
Please add `//! member variable docs` for these private members

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -20,18 +20,20 @@
 
 #include "openPMD/IO/ADIOS/ADIOS2IOHandler.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <set>
+#include <string>
```
Stdlib includes last please

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -1315,41 +1464,55 @@ namespace detail
+    openPMD::AccessType at,
+    MPI_Comm comm,
+    nlohmann::json options )
+    : AbstractIOHandler( std::move( path ), at, comm )
+    , m_impl{ this, comm, std::move( options ) }
```
```suggestion
    : AbstractIOHandler( std::move( path ), at, comm ),
      m_impl{ this, comm, std::move( options ) }
```

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -1315,41 +1464,55 @@ namespace detail
+    std::string path,
+    AccessType at,
+    nlohmann::json options )
+    : AbstractIOHandler( std::move( path ), at )
+    , m_impl{ this, std::move( options ) }
```
```suggestion
    : AbstractIOHandler( std::move( path ), at ),
      m_impl{ this, std::move( options ) }
```

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -1315,41 +1464,55 @@ namespace detail
+ADIOS2IOHandler::ADIOS2IOHandler(
+    std::string path,
+    AccessType at,
+    MPI_Comm comm
+    , nlohmann::json
```
```suggestion
    MPI_Comm comm,
    nlohmann::json
```

ax3l commented on src/auxiliary/JSON.cpp:
```diff
@@ -0,0 +1,94 @@
+#include "openPMD/auxiliary/JSON.hpp"
+
+#include <vector>
+
+#include "openPMD/config.hpp"
```
```suggestion
#include "openPMD/auxiliary/JSON.hpp"
#include "openPMD/config.hpp"

#include <vector>
```

ax3l replied:
and missing copyright header

ax3l commented on src/auxiliary/JSON.cpp:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    TracingJSON::TracingJSON( nlohmann::json originalJSON )
+        : m_originalJSON(
+              std::make_shared< nlohmann::json >( std::move( originalJSON ) ) )
+        , m_shadow( std::make_shared< nlohmann::json >() )
+        , m_positionInOriginal( &*m_originalJSON )
+        , m_positionInShadow( &*m_shadow )
```
```suggestion
        : m_originalJSON(
              std::make_shared< nlohmann::json >( std::move( originalJSON ) ) ),
          m_shadow( std::make_shared< nlohmann::json >() ),
          m_positionInOriginal( &*m_originalJSON ),
          m_positionInShadow( &*m_shadow )
```

ax3l commented on src/auxiliary/JSON.cpp:
```diff
@@ -0,0 +1,94 @@
+    void
+    TracingJSON::invertShadow(
+        nlohmann::json & result,
+        nlohmann::json & shadow )
```
is `shadow` a `const &`?

franzpoeschel replied:
Yes, this method mustn't modify `shadow`. I added the `const`.

ax3l commented on src/auxiliary/JSON.cpp:
```diff
@@ -0,0 +1,94 @@
+        nlohmann::json * positionInOriginal,
+        nlohmann::json * positionInShadow,
+        bool trace )
+        : m_originalJSON( std::move( originalJSON ) )
+        , m_shadow( std::move( shadow ) )
+        , m_positionInOriginal( positionInOriginal )
+        , m_positionInShadow( positionInShadow )
+        , m_trace( trace )
+    {
```
```suggestion
        : m_originalJSON( std::move( originalJSON ) ),
          m_shadow( std::move( shadow ) ),
          m_positionInOriginal( positionInOriginal ),
          m_positionInShadow( positionInShadow ),
          m_trace( trace )
    {
```

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -3,23 +3,23 @@
 #include <tuple>
+#include <vector>
+
+#include "openPMD/auxiliary/Filesystem.hpp"
+#include "openPMD/auxiliary/Environment.hpp"
+#include "openPMD/openPMD.hpp"
+#include <catch2/catch.hpp>
```
include order is off

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -2320,3 +2320,119 @@ TEST_CASE( "no_serial_adios1", "[serial][adios]")
+  }
+}
+)END";
+    auto read = []( std::string const & filename, std::string const & config ) {
```
```suggestion
    auto const read = []( std::string const & filename, std::string const & config ) {
```

franzpoeschel committed:
Apply suggestions from code review

Co-Authored-By: Axel Huebl <axel.huebl@plasma.ninja>
